tests-e2e/state_updates_test.go

```diff
@@ -0,0 +1,251 @@
+			roomID: {
+				m.MatchJoinCount(2),
+				m.MatchRoomName(bob.Localpart),
+				// despite asking for 5 timeline events, we only get 1 on the initial sync so prev_batch is correct
+				MatchRoomTimeline([]Event{
+					{
+						ID: firstTopicEventID,
+					},
+				}),
+			},
+		},
+	))
+
+	// alice leaves the room => no proxy user in the room anymore!
+	alice.MustLeaveRoom(t, roomID)
+	res = alice.SlidingSyncUntilEvent(t, res.Pos, sync3.Request{}, roomID, Event{
+		Content: map[string]interface{}{
+			"membership": "leave",
+		},
+		Type:     "m.room.member",
+		StateKey: &alice.UserID,
+		Sender:   alice.UserID,
+	})
+
+	// now a few things happen:
+	// - charlie joins the room (ensures joins are reflected)
+	// - bob leaves the room (ensures leaves are reflected)
+	// - doris is invited to the room (ensures invites are reflected)
+	// - the topic is changed (ensures state updates)
+	// - the room avatar is set (ensure avatar field updates and new state is added)
+	// - 50 messages are sent (forcing a gappy sync)
+	charlie.MustJoinRoom(t, roomID, []string{"hs1"})
+	bob.MustInviteRoom(t, roomID, doris.UserID)
+	newTopic := "New Topic Name"
+	newAvatar := "mxc://example.org/JWEIFJgwEIhweiWJE"
+	sendTopic(t, bob, roomID, newTopic)
+	sendAvatar(t, bob, roomID, newAvatar)
+	bob.MustLeaveRoom(t, roomID)
+	messageEventIDs := make([]string, 0, 50)
+	for i := 0; i < 50; i++ {
+		messageEventIDs = append(messageEventIDs, sendMessage(t, charlie, roomID, fmt.Sprintf("message %d", i)))
+	}
+
+	// alice rejoins the room.
+	alice.MustJoinRoom(t, roomID, []string{"hs1"})
+
+	// we should get a 400 expired connection
+	start := time.Now()
+	for {
+		httpRes := alice.DoSlidingSync(t, sync3.Request{}, WithPos(res.Pos))
+		if httpRes.StatusCode == 400 {
+			t.Logf("connection expired")
+			break
+		}
+		if time.Since(start) > 5*time.Second {
+			t.Fatalf("did not get expired connection on rejoin")
+		}
+		body := client.ParseJSON(t, httpRes)
+		if err := json.Unmarshal(body, &res); err != nil {
+			t.Fatalf("failed to unmarshal response: %v", err)
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	aliceJoin := Event{
+		Content: map[string]interface{}{
+			"displayname": alice.Localpart,
+			"membership":  "join",
+		},
+		Type:     "m.room.member",
+		StateKey: &alice.UserID,
+		Sender:   alice.UserID,
+	}
+	// ensure state is correct: new connection
+	res = alice.SlidingSyncUntilEvent(t, "", sync3.Request{
+		Lists: map[string]sync3.RequestList{
+			"a": {
+				RoomSubscription: sync3.RoomSubscription{
+					TimelineLimit: 5,
+					RequiredState: [][2]string{{"*", "*"}}, // all state
+				},
+				Ranges:         sync3.SliceRanges{{0, 20}},
+				BumpEventTypes: []string{"m.room.message"},
+			},
+		},
+	}, roomID, aliceJoin)
+	m.MatchResponse(t, res, m.MatchRoomSubscriptionsStrict(map[string][]m.RoomMatcher{
+		roomID: {
+			m.MatchInviteCount(1), // doris
+			m.MatchJoinCount(2),   // alice and charlie
+			MatchRoomTimeline([]Event{aliceJoin}),
```

Comment:
Why not Charlie's messages?

Reply:
Because alice's join is the latest event, and the proxy always applies "joined" history visibility, relying on the client hitting /messages for scrollback, hence it's bounded by her join.